conda/common/configuration.py

```diff
@@ -1609,3 +1609,69 @@ def typify_parameter(self, parameter_name, value, source):
 
     def get_descriptions(self):
         raise NotImplementedError()
+
+
+def get_plugin_config_data(
+    data: dict[Path, dict[str, RawParameter]],
+) -> dict[Path, dict[str, RawParameter]]:
+    """
+    This is used to move everything under the key "plugins" from the provided dictionary
+    to the top level of the returned dictionary. The returned dictionary is then passed
+    to :class:`PluginConfig`.
+    """
+    new_data = {}
+
+    for source, config in data.items():
+        if plugin_data := config.get("plugins"):
+            for param_name, raw_param in plugin_data.value(None).items():
+                new_data[source] = {param_name: raw_param}
+
+        if source == "envvars":
+            for env_var, raw_param in config.items():
+                if env_var.startswith("plugins_"):
+                    _, param_name = env_var.split("plugins_")
+                    new_data[source] = {param_name: raw_param}
+
+    return new_data
+
+
+class PluginConfig(metaclass=ConfigurationType):
+    """
+    Class used to hold settings for conda plugins.
+
+    This object created by this class should only be accessed via
+    :class:`conda.base.context.Context.plugins`.
+
+    This class is updated via the :meth:`add_setting` method which
+    is called in :meth:`conda.plugins.manager.CondaPluginManager.load_settings`.
+    We do this because ``CondaPluginManager`` has access to all registered plugin hooks.
+    """
+
+    def __init__(self, data):
+        self._cache_ = {}
+        self.raw_data = get_plugin_config_data(data)
+
+    @classmethod
+    def add_setting(cls, name: str, loader: ParameterLoader) -> None:
```

Comment:
I don't think that will be possible because we need the `PluginManager` to be able to modify the settings for plugins. This method provides the interface for that and making it private would block that.

I also think that exposing these methods is not ideal. One potential way around this is to dynamically generate the type via a function that the `PluginManager` could use. Here's what that might look like:

```python

class PluginConfigBase(metaclass=ConfigurationType):
    """This is included to be able to define the metaclass"""
    pass

def create_plugin_config_type(plugin_settings: list[CondaSetting]) -> type:
    def __init__(self, data):
        self._cache_ = {}
        self.raw_data = get_plugin_config_data(data)

    plugin_config_type = type("PluginConfig", (PluginConfigBase), {
        "__init__": __init__,
    })

    for plugin_setting in plugin_settings:
        setattr(plugin_config_type, plugin.name, plugin.loader)

    return plugin_config_type
```

Using this method means no more special `add_settings` or `remove_all_settings` methods because the `PluginConfig` type will now be seen as "immutable" If any changes need to be made, we simply generate a new type and throw away the old one. It does look a little weird though because it involves dynamic type generation.

I think I like this method more though. Would be interested to hear your thoughts.

Reply:
Just adding that I don't agree that dynamically creating classes is superior to the class methods, just that the class methods should not be bunched up with the configuration system like that. Just move them into a separate in-between tool class like `PluginConfiguration` with `add` and `remove` steps, or something, that allows us glue the config and plugin systems together.